llvm/lib/DWARFCFIChecker/Registers.h

```diff
@@ -0,0 +1,68 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains helper functions to find and list registers that are
+/// tracked by the unwinding information checker.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_DWARFCFICHECKER_REGISTERS_H
+#define LLVM_DWARFCFICHECKER_REGISTERS_H
+
+#include "llvm/MC/MCRegister.h"
+#include "llvm/MC/MCRegisterInfo.h"
+#include <iterator>
+
+namespace llvm {
+
+/// This analysis only keeps track and cares about super registers, not the
+/// subregisters. All reads from/writes to subregisters are considered the
+/// same operation to super registers.
+inline bool isSuperReg(const MCRegisterInfo *MCRI, MCPhysReg Reg) {
+  return MCRI->superregs(Reg).empty();
+}
+
+inline SmallVector<MCPhysReg> getSuperRegs(const MCRegisterInfo *MCRI) {
+  SmallVector<MCPhysReg> SuperRegs;
+  for (auto &&RegClass : MCRI->regclasses()) {
+    for (unsigned I = 0; I < RegClass.getNumRegs(); I++) {
+      MCPhysReg Reg = RegClass.getRegister(I);
+      if (isSuperReg(MCRI, Reg))
+        SuperRegs.push_back(Reg);
+    }
+  }
+
+  sort(SuperRegs.begin(), SuperRegs.end());
+  SuperRegs.resize(std::distance(
+      SuperRegs.begin(), std::unique(SuperRegs.begin(), SuperRegs.end())));
+  return SuperRegs;
+}
+
+inline SmallVector<MCPhysReg> getTrackingRegs(const MCRegisterInfo *MCRI) {
+  SmallVector<MCPhysReg> TrackingRegs;
+  for (auto Reg : getSuperRegs(MCRI))
+    if (!MCRI->isArtificial(Reg) && !MCRI->isConstant(Reg))
+      TrackingRegs.push_back(Reg);
+  return TrackingRegs;
+}
+
+inline MCPhysReg getSuperReg(const MCRegisterInfo *MCRI, MCPhysReg Reg) {
+  if (isSuperReg(MCRI, Reg))
+    return Reg;
+  for (auto SuperReg : MCRI->superregs(Reg)) {
```

Comment:
Some inconsistency with bracing in this patch - like this loop has braces, but a very similar loop at line 49 doesn't... I think technically according to the style guide, both should omit braces, but I could go either way.

(there's also some other rather long-bodied `for/if{}` without the braces on the for which get a bit of a stretch - but, yeah, the rules are pretty vague)